### DISCUSSION TITLE
Best way to save data

### DISCUSSION THREAD

**Discussion by bajrangCoder** [Microsoft/node-pty#634] [Q&A] [answered]
Is there any better way to save terminal content automatically on new data?
As i have tried on 
```js
pty.onData((data) => {
  session[pid].terminalData += data;
  ws.send(data);
});
```

But as expected it doesn't works because pty sends terminal sequence

**Comment by Tyriar** [comment]
This is more of a general JS question, you could append it to a string like that but it's not very efficient. Alternatively you could append lines to a string array which is a little better, or persist to a file, etc.

**Reply by bajrangCoder** [reply]
Hmm, but pty sends many useless data which aren't useless for terminal because that will be parsed and then shown on xtermjs terminal but in string or file it will save every data .
For eg:
If frontend sends `clear` command then pty sends a clear terminal but while saving it will just save it as a string and one more case in which this fails is :  when user uses zsh Autocompletion plugin then pty sends the same prompt many times corresponding to Autocompletion

**Reply by Tyriar** [reply, marked as answer]
You can feed it into xterm.js (or xterm-headless on node) and then use the serialize addon. I think that would give you what you're after.

This is how terminal restoration works in vscode when reopening windows.

**Comment by bajrangCoder** [comment]
> You can feed it into xterm.js (or xterm-headless on node) and then use the serialize addon. I think that would give you what you're after.
> 
> This is how terminal restoration works in vscode when reopening windows.



> You can feed it into xterm.js (or xterm-headless on node) and then use the serialize addon. I think that would give you what you're after.
> 
> This is how terminal restoration works in vscode when reopening windows.

Thanks 👍
This is what I wanted.
